### PR TITLE
Add Unsafe Byte/Short atomic intrinsics in AArch64

### PR DESCRIPTION
**Pull request by debayang**
Adding support for the below Unsafe intrinsic operations for AArch64 architecture in Graal.

Unsafe.getAndSet(Byte/Short)
Unsafe.getAndAdd(Byte/Short)
Unsafe.compareAndExchange(Byte/Short)
Unsafe.compareAndSet(Byte/Short)

### CODE REVIEW

**Comment by sanzinger** [comment]
So in general, extending the bits during code generation is not recommended. Instead we should generate the sign/zero extend nodes as input for the `AtomicReadAndWriteNode` and `AtomicReadAndAddNode` in the higher level graphs.

**Comment by sanzinger** [comment]
So overall, we have to be careful, where we silently extend the bit widths.

If we want to support sub-wort atomic intrinsifications, we have to distinguish between the store kind and the actual bitwidth of the operation (such as `getAndAdd`).

Here are my thoughts on each subword intrinsification class:
### `getAndSet`
LSE: `AArch64MacroAssembler.swp` supports 8, 16, 32 and 64 bit values.
Non-LSE: `ldaxr` and `stlxr` supports 8, 16, 32, and 64 bit values.

### `getAndAdd`
LSE: Instruction `ldadd` supports only 32 and 62 bit. Cannot be used, overflow in the addition can destroy neighboring bits in memory.
Non-LSE: Instruction `ldaxr` and `stlxr` support 8, 16, 32 and 64 bit. The `add` instruction does only support 32 and 64 bit. Overflows do not harm here, since `stlxr` only stores the lower 8 or 16 bits anyway.

### `compareAndExchange` (`CompareAndSwapOp`)
LSE: Instruction `cas` only supports 32 and 64 bit.
Non-LSE: `ldaxr` and `stlxr` work with 8, 16, 32 and 64 bits, we only have to make sure, that the compare `gpCompare` works correct.

File: compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64AtomicMove.java [hunk 1]
```diff
@@ -73,17 +75,17 @@ public CompareAndSwapOp(AllocatableValue result, Value expectedValue, Allocatabl
...
 
             Register address = asRegister(addressValue);
             Register result = asRegister(resultValue);
             Register newVal = asRegister(newValue);
             if (AArch64LIRFlagsVersioned.useLSE(masm.target.arch)) {
                 Register expected = asRegister(expectedValue);
-                masm.mov(size, result, expected);
-                masm.cas(size, result, newVal, address, true /* acquire */, true /* release */);
+                masm.mov(regsize, result, expected);
+                masm.cas(regsize, result, newVal, address, true /* acquire */, true /* release */);
```
[the remark this answers is not in the document]
Comment (sanzinger) [reply compiler/src/org.graalvm.compiler.lir.aarch64/src/org/graalvm/compiler/lir/aarch64/AArch64AtomicMove.java]: I have to correct this. AArch64 has CAS for 16 bit ([cash](http://www.keil.com/support/man/docs/armclang_asm/armclang_asm_aoa1476202790353.htm)) and 8 bit ([casb](http://www.keil.com/support/man/docs/armclang_asm/armclang_asm_gmf1476202790003.htm)). So what is actually missing is the implementation in the `AArch64Assembler` and the proper selection of it.

**Comment by a74nh** [comment]
@debayang Do you have any plans to fix this patch? If not then we can look at reimplementing this.

**Comment by teshull** [comment]
It looks like a new version of this PR has been introduced [here](https://github.com/oracle/graal/pull/2763) Therefore, I am closing this PR.

**Pull request closed without merging**